### PR TITLE
Implement short-circuiting for assertions

### PR DESCRIPTION
**Pull request by Schaeff**
currently

```
if c then foo() else bar() fi
```

would always panic if `foo()` panics (even if `c` in false)

This is because we inline foo(), and conditionally return its result, but if a panic happened in the statements of foo, it would happen even if this branch isn't executed.

With this change, panics in each branch lead to the whole execution panicking only if the given branch is "executed", ie if `c`is true.

Both branches are still technically being executed (and paid for in terms of constraints), but the behavior is closer to what would expect.

Todo:
- [x] minimize constraints: next PR
- [x] try to reduce what we can before flattening to benefit from boolean propagation: next PR
- [x] tests nested conditionals
- [x] fix indentation on display implementation 
- [x] put behind flag?

### CODE REVIEW

File: zokrates_core/src/typed_absy/mod.rs [hunk 1]
```diff
@@ -714,9 +715,26 @@ impl<'ast, T> MultiTyped<'ast, T> for TypedExpressionList<'ast, T> {
     }
 }
 
+#[derive(Clone, PartialEq, Debug, Hash, Eq)]
+// a block expression which returns an `E`
```
Comment (dark64) [inline zokrates_core/src/typed_absy/mod.rs]: For the sake of consistency, I think this comment should go above the attribute

**Comment by dark64** [comment]
Compile config should be updated in the `zokrates_js` type definition file as well: https://github.com/Zokrates/ZoKrates/blob/conditional-panics/zokrates_js/index.d.ts#L14

**Comment by Schaeff** [comment]
> Compile config should be updated in the `zokrates_js` type definition file as well: https://github.com/Zokrates/ZoKrates/blob/conditional-panics/zokrates_js/index.d.ts#L14

changed

**Review by dark64 — APPROVED** [review]
LGTM!

**Pull request merged**